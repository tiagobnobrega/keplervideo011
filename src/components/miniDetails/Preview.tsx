import { isEqual } from 'lodash';
import React from 'react';
import { Image, StyleSheet, View } from 'react-native';
import { scaleUxToDp } from '../../utils/pixelUtils';
const LeftGradient = require('../../assets/mini_details_left_gradient.png');
const BottomGradient = require('../../assets/mini_details_bottom_gradient.png');

interface PreviewProps {
  posterUrl?: string;
}

const Preview = ({ posterUrl }: PreviewProps) => {
  return (
    <View style={styles.previewContent}>
      {posterUrl && (
        <Image
          testID={'poster-image'}
          style={styles.posterImage}
          source={{ uri: posterUrl }}
          resizeMode={'contain'}
        />
      )}
      <GradientOverlay />
    </View>
  );
};

const GradientOverlay = () => {
  return (
    <View>
      <Image
        style={styles.gradient}
        resizeMode={'contain'}
        source={LeftGradient}
        testID="left-gradient"
      />
      <Image
        style={styles.gradient}
        resizeMode={'contain'}
        source={BottomGradient}
        testID="bottom-gradient"
      />
    </View>
  );
};

const posterUrlDataUnchanged = (
  prevProps: PreviewProps,
  nextProps: PreviewProps,
) => {
  return isEqual(prevProps.posterUrl, nextProps.posterUrl);
};

export default React.memo(Preview, posterUrlDataUnchanged);

const styles = StyleSheet.create({
  gradient: {
    width: scaleUxToDp(1236),
    height: scaleUxToDp(696),
    position: 'absolute',
  },
  previewContent: {
    left: scaleUxToDp(682),
    position: 'absolute',
  },
  posterImage: {
    width: scaleUxToDp(1236),
    height: scaleUxToDp(696),
    position: 'absolute',
  },
});
