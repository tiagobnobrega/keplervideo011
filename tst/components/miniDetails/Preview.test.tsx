import { render } from '@testing-library/react-native';
import { isEqual } from 'lodash';
import React from 'react';
import Preview from '../../../src/components/miniDetails/Preview';

jest.mock('lodash', () => ({
  isEqual: jest.fn(),
}));

describe('Preview renders correctly', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    (jest.mocked(isEqual) as jest.Mock).mockImplementation((a, b) => {
      return JSON.stringify(a) === JSON.stringify(b);
    });
  });

  it('renders correctly with a posterUrl', () => {
    const posterUrl = 'http://example.com/poster.jpg';
    const { toJSON, getByTestId } = render(<Preview posterUrl={posterUrl} />);
    expect(toJSON()).toMatchSnapshot();
    expect(getByTestId('poster-image')).toBeTruthy();
  });

  it('renders GradientOverlay correctly', () => {
    const { getByTestId } = render(<Preview />);
    expect(getByTestId('left-gradient')).toBeTruthy();
    expect(getByTestId('bottom-gradient')).toBeTruthy();
  });
});

describe('React.memo behavior for Preview', () => {
  const mockPosterUrl = 'http://example.com/poster.jpg';

  it('does not re-render when posterUrl is unchanged', () => {
    const { rerender } = render(<Preview posterUrl={mockPosterUrl} />);

    rerender(<Preview posterUrl={mockPosterUrl} />);
    expect(isEqual).toHaveBeenCalledWith(mockPosterUrl, mockPosterUrl);
    expect(isEqual).toHaveBeenCalledTimes(1);
  });

  it('re-renders when posterUrl changes', () => {
    const { rerender } = render(<Preview posterUrl={mockPosterUrl} />);

    const newPosterUrl = 'http://example.com/new-poster.jpg';

    rerender(<Preview posterUrl={newPosterUrl} />);
    expect(isEqual).toHaveBeenCalledWith(mockPosterUrl, newPosterUrl);
  });
});
