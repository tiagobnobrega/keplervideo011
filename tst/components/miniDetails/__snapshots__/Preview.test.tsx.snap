// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preview renders correctly renders correctly with a posterUrl 1`] = `
<View
  style={
    Object {
      "left": 682,
      "position": "absolute",
    }
  }
>
  <Image
    resizeMode="contain"
    source={
      Object {
        "uri": "http://example.com/poster.jpg",
      }
    }
    style={
      Object {
        "height": 696,
        "position": "absolute",
        "width": 1236,
      }
    }
    testID="poster-image"
  />
  <View>
    <Image
      resizeMode="contain"
      source={
        Object {
          "testUri": "../../../src/assets/mini_details_left_gradient.png",
        }
      }
      style={
        Object {
          "height": 696,
          "position": "absolute",
          "width": 1236,
        }
      }
      testID="left-gradient"
    />
    <Image
      resizeMode="contain"
      source={
        Object {
          "testUri": "../../../src/assets/mini_details_bottom_gradient.png",
        }
      }
      style={
        Object {
          "height": 696,
          "position": "absolute",
          "width": 1236,
        }
      }
      testID="bottom-gradient"
    />
  </View>
</View>
`;
